### PR TITLE
Cleanup and prep pivot table backing classes

### PR DESCRIPTION
**Pull request by laum18**

### CODE REVIEW

File: EPPlus/Table/PivotTable/SharedItemsCollection.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Xml;
+using OfficeOpenXml.Utils;
+
+namespace OfficeOpenXml.Table.PivotTable
+{
+	public class SharedItemsCollection : XmlHelper
```
Comment (EvanSchallerer) [inline EPPlus/Table/PivotTable/SharedItemsCollection.cs]: Comment

File: EPPlus/Table/PivotTable/CacheFieldNode.cs [hunk 2]
```diff
@@ -76,49 +77,38 @@ public IReadOnlyList<CacheFieldItem> Items
...
 		/// </summary>
+		/// <param name="type">The type of the target value.</param>
 		/// <param name="value">The target value in the list.</param>
 		/// <returns>The index of the value in the list.</returns>
-		public int GetSharedItemIndex(string value)
+		public int GetSharedItemIndex(PivotCacheRecordType type, object value)
 		{
-			for (int i = 0; i < this.Items.Count; i++)
+			string stringValue = ConvertUtil.ConvertObjectToXmlAttributeString(value);
+			for (int i = 0; i < this.SharedItems.Count; i++)
```
Comment (shawncrabtree) [inline EPPlus/Table/PivotTable/CacheFieldNode.cs]: Use the FindIndex method.
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.findindex?redirectedfrom=MSDN&view=netframework-4.7.2#System_Collections_Generic_List_1_FindIndex_System_Predicate__0__

File: EPPlusTest/Table/PivotTable/CacheItemTest.cs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+		[TestMethod]
+		[ExpectedException(typeof(InvalidOperationException))]
+		public void GetObectTypeInvalidType()
+		{
+			CacheItem.GetObjectType(new List<int>());
+		}
+		#endregion
+
+		#region UpdateValue Tests
+		//[TestMethod]
```
Comment (EvanSchallerer) [inline EPPlusTest/Table/PivotTable/CacheItemTest.cs]: We should put a comment with the task # that we are going to put these tests in with, or just remove them for this pr

File: EPPlus/Table/PivotTable/CacheRecordNode.cs [hunk 4]
```diff
@@ -105,7 +107,98 @@ public CacheRecordNode(XmlNode node, XmlNamespaceManager namespaceManager)
...
+			if (row == null)
+				throw new ArgumentNullException(nameof(row));
+			if (cacheDefinition == null)
+				throw new ArgumentNullException(nameof(cacheDefinition));
+			this.NameSpaceManager = namespaceManager;
+			var recordNode = parentNode.OwnerDocument.CreateElement("d:r");
+			int col = 0;
+			foreach (var value in row)
+			{
+				var type = CacheItem.GetObjectType(value);
```
Comment (shawncrabtree) [inline EPPlus/Table/PivotTable/CacheRecordNode.cs]: this.CacheItem?

File: EPPlus/Table/PivotTable/SharedItemsCollection.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
```
Comment (zpfaltersack) [inline EPPlus/Table/PivotTable/SharedItemsCollection.cs]: license

File: EPPlus/Table/PivotTable/SharedItemsCollection.cs (hunk 1 above)
Comment (zpfaltersack) [inline EPPlus/Table/PivotTable/SharedItemsCollection.cs]: comments

File: EPPlus/Table/PivotTable/SharedItemsCollection.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		public int Count
+		{
+			get { return base.GetXmlNodeIntNull("@count") ?? 0; }
+			private set { base.SetXmlNodeString("@count", value.ToString()); }
+		}
+
+		/// <summary>
+		/// Gets a readonly list of the items in this <see cref="CacheFieldNode"/>.
+		/// </summary>
+		public IReadOnlyList<CacheItem> Items
```
Comment (shawncrabtree) [inline EPPlus/Table/PivotTable/SharedItemsCollection.cs]: Arrow function

Comment (shawncrabtree) [reply EPPlus/Table/PivotTable/SharedItemsCollection.cs]: Actually just get rid of the class variable and make it a {get; private set;}

Comment (shawncrabtree) [reply EPPlus/Table/PivotTable/SharedItemsCollection.cs]: Actually just a {get;}

File: EPPlusTest/Table/PivotTable/CacheItemTest.cs (hunk 3 above)
Comment (zpfaltersack) [inline EPPlusTest/Table/PivotTable/CacheItemTest.cs]: delete

File: EPPlusTest/Table/PivotTable/CacheItemTest.cs [hunk 7]
```diff
@@ -0,0 +1,320 @@
+﻿using System;
```
Comment (zpfaltersack) [inline EPPlusTest/Table/PivotTable/CacheItemTest.cs]: might as well license it

File: EPPlusTest/Table/PivotTable/SharedItemsCollectionTest.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
```
Comment (zpfaltersack) [inline EPPlusTest/Table/PivotTable/SharedItemsCollectionTest.cs]: license

File: EPPlus/Table/PivotTable/CacheFieldNode.cs (hunk 2 above)
Comment (laum18) [reply EPPlus/Table/PivotTable/CacheFieldNode.cs]: IReadOnlyList doesn't have this method.

File: EPPlus/Table/PivotTable/CacheRecordNode.cs (hunk 4 above)
Comment (laum18) [reply EPPlus/Table/PivotTable/CacheRecordNode.cs]: CacheItem is a type, GetObjectType is a static method.

File: EPPlus/Table/PivotTable/SharedItemsCollection.cs (hunk 6 above)
Comment (EvanSchallerer) [reply EPPlus/Table/PivotTable/SharedItemsCollection.cs]: That won't work. The class variable is there to allow us to modify the collection by keep it from being modified by outside classes.

**Pull request merged**